### PR TITLE
drivers: dynamic power handling of entropy_src

### LINKED ISSUE #2868: [RFC] Dynamic power handling of entropy source
# Dynamic power handling of entropy source
This RFC outlines entropy source handling for better power consumptions within Caliptra integrated devices.

## Scope
Document update: a short text about `entropy_src` handling might be added to the Caliptra RT specification.
Security posture: This change will be impacted by NIST SP 800-90A. The implementation will be done complying NIST SP 800-90A requirement regarding entropy source.
Memory consumption: This change will slightly increase the code size, but it will be negligible since they are only few lines for condition checking and changing HW configurations.
Runtime: Execution time will be slightly increase as every trng generate call will be led by reseed counter check, but it will be negligible.

## Rationale
Power consumption is one of important metrics for any type of device. This change will improve power profile without compromising anything. The OpenTitan CSRNG is designed to support such functionality to improve power profile of devices (https://opentitan.org/book/hw/ip/csrng/doc/programmers_guide.html#running-csrng-with-entropy_src-disabled).
The Caliptra core is currently configuring `RESEED_INTERVAL` to be `0xFFFFFFFF`, which means that once the CSRNG needs to be reseeded after 2<sup>32</sup> requests. 

When `entropy_src` is enabled, it keeps generating fresh entropy with the underlying noise source whether it is used or not. So, this computation is completely unnecessary, and this is the point where we can improve the power profile without losing anything. Moreover, it will result in relaxing hardware requirements on the noise source to be connected to `entropy_src` since the noise source will not be fully loaded anymore with this change.

## Implementation Tradeoffs
The OpenTitan CSRNG provides two options:
1. regular, non-deterministic mode; and
2. fully deterministic mode.

Since the Caliptra internal TRNG is instantiated in non-deterministic mode, it would be better to take the first option as we can at least avoid using all-zero seed which can be happened with fully deterministic mode in the worst case scenario.

In order to simplify implementation, we would like to touch `entropy_src` in the boot flow. 
With the first option, we are foreseeing the following changes within the Caliptra RT implementation:
1. Once the Caliptra RT is operational, the Caliptra RT disables `entropy_src` by using `MODULE_ENABLE` register of `entropy_src`.
2. On every CSRNG `Generate` command call, the Caliptra RT checks if `RESEED_COUNTER` is *close enough* to `RESEED_INTERVAL`. If so, the Caliptra RT enables `entropy_src`.
   - depending on the choice of the threshold to be considered as *"close enough"* `entropy_src` health test will not come as an extra overhead for reseeding.
4. On every CSRNG `Generate` command call, the Caliptra RT checks if RESEED_COUNTER is *close enough* to zero. If so, the Caliptra RT disables `entropy_src`.
6. On every CSRNG `Reseed` command call, the Caliptra RT first enables `entropy_src`, then issues the `Reseed` command, and finally disables `entropy_src` again. 
Since `RESEED_COUNTER` and `RESEED_INTERVAL` are only internally available, this change should be included within the Caliptra RT.

## Implementation Timeline
A realistic estimate for completion. May include multiple milestones if a large change requires many Pull Requests.
- The implementation can be done in one week, and it will take another week for testing to cover as many edge cases as possible.

Which MAJOR.MINOR release is this intended for inclusion?
- We are planning to include this change in 2.1 release.

## Test Plan
In order to check our expected behavior on CSRNG is correct, we will test the following cases:
1. CSRNG behavior when the entropy source is disabled;
   - when it reaches to `RESEED_INTERVAL` and `Reseed` command is explictly called;
2. CSRNG behavior when `RESEED_COUNTER` reaches to `RESEED_INTERVAL`; and
3. `RESEED_COUNTER` behavior after `Reseed` command call.

Then, these are the test cases to check our implementation:
1. CSRNG `Generate` command call when `RESEED_COUNTER` is *close enough* to `RESEED_INTERVAL`;
2. CSRNG `Generate` command call when `RESEED_COUNTER` is close to zero;
3. CSRNG `Generate` command call when `RESEED_INTERVAL` changed to a value smaller than `RESEED_COUNTER`;
4. CSRNG `Generate` command call when `RESEED_INTERVAL` changed to a value after `entropy_src` is enabled;
5. CSRNG behavior when `RESEED_COUNTER` exceeds `RESEED_INTERVAL` while the entropy source is not enabled;
6. CSRNG behavior when `RESEED_COUNTER` exceeds `RESEED_INTERVAL` while the entropy source is failed for health test;

While implementing, we will develop more test cases if needed.

## Maintenance
This change will be maintained by the Samsung electronics memory solution security team. (contact: @gwangbaeCHOI)

### PR DESCRIPTION
**Pull request by GwangbaeCHOI**
Keep entropy_src disabled outside of CSRNG Reseed windows to reduce power consumption, per RFC #2868. Caliptra RT leaves entropy_src MODULE_ENABLE=False by default and only enables it around a Reseed, with a headroom so the health-test startup is ready when the Reseed fires. RESEED_INTERVAL stays at the hardware default — the saving comes from gating entropy_src, not from shortening the interval.

reseed() waits for the CSRNG main SM to return to MainSmIdle before returning, since send_command returns when the Reseed is accepted into the staging FIFO, not when the CSRNG finishes executing it.

generate12/generate16 surface a runtime health-check failure via check_runtime_health, which calls check_for_alert_state only while entropy_src is enabled — calling it with entropy_src disabled would hang waiting for a main-sm state that never comes.

Fixes #2868

### CODE REVIEW

File: drivers/src/csrng.rs [hunk 1]
```diff
@@ -185,7 +184,7 @@ impl Csrng {
     /// }
     /// ```
     pub fn generate12(&mut self) -> CaliptraResult<[u32; 12]> {
-        check_for_alert_state(self.entropy_src.regs())?;
```
Comment (GwangbaeCHOI) [inline drivers/src/csrng.rs:188]: `generate12` only uses the CSRNG and the CSRNG doesn't interact with the entropy source to generate random bits. Checking the entropy source state is not necessary.

File: drivers/src/csrng.rs [hunk 2]
```diff
@@ -235,7 +287,7 @@ impl Csrng {
     ///
     /// Returns an error if the internal generate command fails.
     pub fn generate16(&mut self) -> CaliptraResult<[u32; 16]> {
-        check_for_alert_state(self.entropy_src.regs())?;
```
Comment (GwangbaeCHOI) [inline drivers/src/csrng.rs:238]: Same with the `generate12` case.
https://github.com/chipsalliance/caliptra-sw/pull/4038/changes#r3670202720

File: runtime/src/main.rs [hunk 3]
```diff
@@ -103,6 +103,10 @@ pub extern "C" fn entry_point() -> ! {
     if !drivers.persistent_data.get().rom.fht.is_valid() {
         handle_fatal_error(caliptra_drivers::CaliptraError::RUNTIME_HANDOFF_FHT_NOT_LOADED.into());
     }
+
+    cprintln!("[rt] Disable entropy source");
+    drivers.trng.disable_entropy_source();
```
Comment (GwangbaeCHOI) [inline runtime/src/main.rs:108]: Disable the entropy source once the RT is loaded. Therefore, disabling the entropy source won't delay the boot time.

File: drivers/src/csrng.rs [hunk 4]
```diff
@@ -222,6 +221,59 @@ impl Csrng {
         }
     }
 
+    /// Number of Generate calls before the reseed interval at which entropy_src
+    /// is pre-enabled, so it is ready by the time the reseed is needed.
+    /// Covers the ~2 ms entropy_src startup at typical generate cadence.
```
Comment (GwangbaeCHOI) [inline drivers/src/csrng.rs:226]: Tests were done on FPGA

File: sw-emulator/lib/periph/src/csrng.rs
Comment (GwangbaeCHOI) [inline sw-emulator/lib/periph/src/csrng.rs:1]: Addding reseeding related logics into the sw emulator

**Comment by clundin25** [comment]
@GwangbaeCHOI can you commit the new ROM hash using `cargo xtask update-frozen-images`?

File: drivers/src/csrng.rs [hunk 5]
```diff
@@ -222,6 +221,59 @@ impl Csrng {
...
+    pub const ENTROPY_SRC_PRE_ENABLE_HEADROOM: u32 = 5;
+
+    /// Disable entropy_src.
+    ///
+    /// Must be called only after any in-flight Reseed has consumed its seed
+    /// from the esfinal FIFO; disabling mid-Reseed clears that FIFO and hangs
+    /// the CSRNG in `MainSmReseedPrep`. Pair with `wait_for_csrng_idle`.
+    pub fn disable_entropy_source(&mut self) {
+        const FALSE: u32 = MultiBitBool::False as u32;
```
Comment (clundin25) [inline drivers/src/csrng.rs]: nit: I think this and `TRUE` are already defined in `with_seed`.

Maybe we should just elevate that scope so we can reuse the constant definition?

File: drivers/src/csrng.rs [hunk 6]
```diff
@@ -222,6 +221,59 @@ impl Csrng {
...
+    /// is pre-enabled, so it is ready by the time the reseed is needed.
+    /// Covers the ~2 ms entropy_src startup at typical generate cadence.
+    pub const ENTROPY_SRC_PRE_ENABLE_HEADROOM: u32 = 5;
+
+    /// Disable entropy_src.
+    ///
+    /// Must be called only after any in-flight Reseed has consumed its seed
+    /// from the esfinal FIFO; disabling mid-Reseed clears that FIFO and hangs
+    /// the CSRNG in `MainSmReseedPrep`. Pair with `wait_for_csrng_idle`.
+    pub fn disable_entropy_source(&mut self) {
```
Comment (clundin25) [inline drivers/src/csrng.rs]: Can we track this state so we don't inadvernately do this? Preferably at compile time since I think this would be a logic bug (Type state pattern may be a good match?)

File: drivers/src/csrng.rs [hunk 7]
```diff
@@ -222,6 +221,59 @@ impl Csrng {
...
+    /// Called before every Generate command. Keeps entropy_src disabled except
+    /// during a short window around each reseed:
+    ///
+    /// Phase 1 — pre-enable (`counter >= interval - HEADROOM`): enable
+    /// entropy_src in the background so its startup health checks overlap with
+    /// the next few Generates.
+    ///
+    /// Phase 2 — reseed (`counter >= interval`): reseed from entropy_src,
+    /// then disable it again. The Generate runs with entropy_src off.
+    fn manage_entropy_src_and_reseed(&mut self) -> CaliptraResult<()> {
```
Comment (clundin25) [inline drivers/src/csrng.rs]: Similarly, I think we could split the Generate commands to a different type that this function returns, guaranteeing we never forget to reseed

**Comment by clundin25** [comment]
@zhalvorsen @vmhatre PTAL

**Comment by GwangbaeCHOI** [comment]
> @GwangbaeCHOI can you commit the new ROM hash using `cargo xtask update-frozen-images`?

I just added it.

File: drivers/src/csrng.rs (hunk 5 above)
Comment (GwangbaeCHOI) [reply drivers/src/csrng.rs]: Fixed

File: drivers/src/csrng.rs (hunk 6 above)
Comment (GwangbaeCHOI) [reply drivers/src/csrng.rs]: Were you thinking of something like Csrng&lt;S> or a different approach? I guess the former would propagate the type parameter into Trng, Drivers, and every ROM/RT function taking &mut Drivers. So, I wonder if this was the approach that you were thinking.

Comment (GwangbaeCHOI) [reply drivers/src/csrng.rs]: For now, changed `pub` to `pub(crate)` on disable_entropy_source and enable_entropy_source so they are not accessible outside the crate.

**Comment by clundin25** [comment]
@GwangbaeCHOI looks like there is a merge conflict, do you mind fixing it?

File: drivers/src/csrng.rs (hunk 6 above)
Comment (clundin25) [reply drivers/src/csrng.rs]: Good point, that sounds good to me. We can consider a future refactor since it will touch so much code.

I think you are right, we would need to change the public API to make it worthwhile

File: drivers/src/csrng.rs (hunk 7 above)
Comment (GwangbaeCHOI) [reply drivers/src/csrng.rs]: Update it.

**Comment by GwangbaeCHOI** [comment]
> @GwangbaeCHOI looks like there is a merge conflict, do you mind fixing it?

I updated it. I guess the conflict is resolved.

**Comment by clundin25** [comment]
PR LGTM but looks like there are some test failures. @GwangbaeCHOI can you take a look?

File: drivers/test-fw/src/bin/csrng_runtime_health_fail_tests.rs [hunk 8]
```diff
@@ -36,10 +36,10 @@ fn test_runtime_health_check_failure() {
...
         .expect("CSRNG should pass startup health testing");
 
-    // First generate should start seeing bad entropy
-    // The emulator will be configured to provide good entropy for startup,
-    // then bad entropy for runtime operations
-    let result = csrng.generate12();
+    // Reseed from entropy_src — this enables entropy_src, consumes from it,
+    // and is where the health check fires. Bad entropy injected by the test
+    // harness will be detected here.
+    let result = csrng.reseed(CsrngSeed::EntropySrc);
```
Comment (GwangbaeCHOI) [inline drivers/test-fw/src/bin/csrng_runtime_health_fail_tests.rs:42]: Changed to `reseed` as `generate12` never cause a entropy source failure.

File: drivers/src/csrng.rs [hunk 9]
```diff
@@ -167,6 +164,33 @@ impl Csrng {
         Ok(result)
     }
 
+    /// Return 4 randomly generated [`u32`]s as a tuple.
+    ///
+    /// The return type is a tuple of 4 `u32`s which fits entirely in registers
+    /// on RV32, guaranteeing no hidden out-pointer and no `memcpy` on the
+    /// return path regardless of optimization level. Safe to call before CFI
+    /// counters are initialized.
+    pub fn generate4(&mut self) -> CaliptraResult<(u32, u32, u32, u32)> {
```
Comment (GwangbaeCHOI) [inline drivers/src/csrng.rs:173]: Added to avoid memcpy before the CFI initialization. Dedicated function is added to avoid using a unsafe block as it was done in old `generate12`.

**Comment by GwangbaeCHOI** [comment]
> PR LGTM but looks like there are some test failures. @GwangbaeCHOI can you take a look?

Failed testcases are addressed.

**Review by clundin25 — APPROVED** [review]
